sycl-fusion/passes/cleanup/Cleanup.cpp

```diff
@@ -0,0 +1,164 @@
+//==---------------------------- Cleanup.cpp -------------------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "Cleanup.h"
+
+#include <algorithm>
+#include <utility>
+
+#include <llvm/ADT/SmallVector.h>
+#include <llvm/IR/Attributes.h>
+#include <llvm/IR/DerivedTypes.h>
+#include <llvm/IR/LLVMContext.h>
+#include <llvm/IR/Metadata.h>
+
+#include "Kernel.h"
+#include "kernel-info/SYCLKernelInfo.h"
+
+using namespace llvm;
+
+static FunctionType *createMaskedFunctionType(const BitVector &Mask,
+                                              const FunctionType *FTy) {
+  SmallVector<Type *> Params;
+  std::transform(Mask.set_bits_begin(), Mask.set_bits_end(),
+                 std::back_inserter(Params),
+                 [&](unsigned I) { return FTy->getParamType(I); });
+  return FunctionType::get(FTy->getReturnType(), Params, false);
+}
+
+static void copyAttributesFrom(const BitVector &Mask, Function *NF,
+                               const Function *F) {
+  // Copies attributes, calling convention and other relevant data.
+  NF->copyAttributesFrom(F);
+  // Drop masked-out attributes.
+  SmallVector<AttributeSet> Attributes;
+  const llvm::AttributeList PAL = NF->getAttributes();
+  std::transform(Mask.set_bits_begin(), Mask.set_bits_end(),
+                 std::back_inserter(Attributes),
+                 [&](unsigned I) { return PAL.getParamAttrs(I); });
+  NF->setAttributes(AttributeList::get(NF->getContext(), PAL.getFnAttrs(),
+                                       PAL.getRetAttrs(), Attributes));
+}
+
+static Function *createMaskedFunction(const BitVector &Mask, Function *F) {
+  // Declare
+  llvm::FunctionType *NFTy =
+      createMaskedFunctionType(Mask, F->getFunctionType());
+  llvm::Function *NF =
+      Function::Create(NFTy, F->getLinkage(), F->getAddressSpace(),
+                       F->getName(), F->getParent());
+  copyAttributesFrom(Mask, NF, F);
+  NF->setComdat(F->getComdat());
+  NF->takeName(F);
+
+  // Copy body
+  NF->getBasicBlockList().splice(NF->begin(), F->getBasicBlockList());
+  {
+    // Transfer uses to new arguments.
+    auto *ArgsIt = NF->arg_begin();
+    auto SetIt = Mask.set_bits_begin();
+    const auto End = Mask.set_bits_end();
+    for (; SetIt != End; ++ArgsIt, ++SetIt) {
+      auto *OldArg = F->getArg(*SetIt);
+      OldArg->replaceAllUsesWith(ArgsIt);
+      ArgsIt->takeName(OldArg);
+    }
+  }
+
+  {
+    // Copy metadata.
+    SmallVector<std::pair<unsigned, MDNode *>> MDs;
```

Comment:
I'm a bit surprised that `SmallVector` doesn't require a second argument, am I missing something? What is the point of using `SmallVector` if we are not pre-allocating some storage on stack?

Reply:
This is actually recommended if there is no well motivated choice for `N`:
> In the absence of a well-motivated choice for the number of inlined elements N, it is recommended to use [SmallVector<T>](https://llvm.org/doxygen/classllvm_1_1SmallVector.html) (that is, omitting the N). This will choose a default number of inlined elements reasonable for allocation on the stack (for example, trying to keep sizeof(SmallVector<T>) around 64 bytes).

From https://llvm.org/doxygen/classllvm_1_1SmallVector.html#details

Reply:
Didn't know there is a default value for `N`, thanks!